File: android/app/src/main/kotlin/org/fgsec/halogen/features/queue/QueueModel.kt
package org.fgsec.halogen.features.queue

import org.fgsec.halogen.core.enqueueMutation
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.launch
import org.fgsec.halogen.components.EpisodeOrderField
import org.fgsec.halogen.components.ListQuery
import org.fgsec.halogen.components.ToastCenter
import org.fgsec.halogen.core.DeviceDownloads
import org.fgsec.halogen.core.DeviceLog
import org.fgsec.halogen.core.FriendlyError
import org.fgsec.halogen.core.HalogenCore
import org.fgsec.halogen.storage.CacheKey
import org.fgsec.halogen.storage.OutboxOp
import org.fgsec.halogen.wire.EpisodeData
import org.fgsec.halogen.wire.OrderDirection
import org.fgsec.halogen.wire.PlaylistData

/// The Queue tab. The queue IS the user's default playlist (`is_default`) — never
/// a separate concept. Local-first: cached rows render before the network answers,
/// and membership mutations apply optimistically then sync via the Outbox.
class QueueModel(private val core: HalogenCore) {
    private val accountStore = core.store


    var queue: PlaylistData? by mutableStateOf(null)
        private set
    var episodes: List<EpisodeData> by mutableStateOf(emptyList())
        private set
    var loaded: Boolean by mutableStateOf(false)
        private set
    var error: String? by mutableStateOf(null)
        private set

    /// Search + facet + order, applied LOCALLY (the queue is a small,
    /// position-ordered list already in memory). Persisted across relaunch
    /// (web use_list_view_state("queue")).
    private val queryState = mutableStateOf(defaultQuery)
    var query: ListQuery
        get() = queryState.value
        set(value) {
            if (value == queryState.value) return
            queryState.value = value
            val store = accountStore
            core.scope.launch { store?.save(value, QUERY_KEY) }
        }

    private var loadedQuery = false
    /// Bumped synchronously by every optimistic mutation. A refresh captures
    /// it before its fetch and discards the response if it moved — a mutation
    /// landing mid-fetch must never be overwritten by the stale server list
    /// (web: cache_playlists applies its membership guard at CACHE time).
    private var mutationEpoch = 0

    suspend fun load() {
        error = null
        if (!loadedQuery) {
            loadedQuery = true
            accountStore?.load<ListQuery>(QUERY_KEY)?.let { queryState.value = it }
        }
        val store = accountStore
        if (store != null) {
            if (queue == null) {
                store.load<PlaylistData>(CacheKey.queueMeta)?.let { queue = it }
            }
            val cachedQueue = queue
            if (cachedQueue != null && episodes.isEmpty()) {
                store.load<List<EpisodeData>>(CacheKey.playlistEpisodes(cachedQueue.id))?.let {
                    episodes = it
                    loaded = true
                }
            }
        }
        refresh()
    }

    suspend fun refresh() {
        // Drain BEFORE pulling (the web's hard-coded order everywhere): the
        // server must reflect optimistic queue membership before we read the
        // playlist back, or this fetch wipes a not-yet-shipped offline add.
        core.outbox?.drain()
        try {
            val fresh = core.defaultPlaylist()
            queue = fresh
            if (fresh != null) {
                accountStore?.save(fresh, CacheKey.queueMeta)
                // Membership ops still queued for this playlist (the drain
                // couldn't ship them): the local list is AHEAD of the server —
                // keep it, don't overwrite screen/cache with stale membership
                // (web: cache_playlists' membership-preservation invariant).
                if (core.outbox?.hasPendingOps(fresh.id) == true) {
                    loaded = true
                    return
                }
                val epoch = mutationEpoch
                val served = core.playlistEpisodes(fresh.id)
                // Guard AGAIN at cache time: a mutation (or its just-enqueued
                // op) that landed while the fetch was in flight outranks the
                // stale server membership.
                if (epoch != mutationEpoch) {
                    loaded = true
                    return
                }
                if (core.outbox?.hasPendingOps(fresh.id) == true) {
                    loaded = true
                    return
                }
                episodes = served
                accountStore?.save(served, CacheKey.playlistEpisodes(fresh.id))
            } else {
                episodes = emptyList()
                accountStore?.remove(CacheKey.queueMeta)
            }
            error = null
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            DeviceLog.warn("queue: refresh failed — ${e::class.simpleName}: ${e.message}")
            if (episodes.isEmpty() && queue == null) error = FriendlyError.message(e)
        }
        loaded = true
    }

    /// "No queue yet" flow: create the default playlist (online-only — the
    /// created id anchors every later offline op). Failure toasts — setting
    /// `error` would replace the whole no-queue screen (and its Create
    /// button) with a full-page error.
    suspend fun createQueue() {
        try {
            core.createPlaylist(name = "Queue", isDefault = true)
            refresh()
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            ToastCenter.error("Couldn't create the queue — ${FriendlyError.message(e)}")
        }
    }

    // ── optimistic membership ops (offline-capable via Outbox) ──────────────

    fun remove(episode: EpisodeData) {
        val queue = queue ?: return
        core.enqueueMutation(OutboxOp.Kind.RemoveFromPlaylist(playlistId = queue.id, episodeId = episode.id)) {
            mutationEpoch += 1
            episodes = episodes.filterNot { it.id == episode.id }
            syncMembership()
        }
    }

    /// `toIndex` is the row's FINAL resting index (the reorder helper's
    /// commit semantics — not SwiftUI's insertion offset).
    fun move(fromIndex: Int, toIndex: Int) {
        val queue = queue ?: return
        val moved = episodes.getOrNull(fromIndex) ?: return
        val finalIndex = toIndex.coerceIn(0, (episodes.size - 1).coerceAtLeast(0))
        core.enqueueMutation(OutboxOp.Kind.MoveInPlaylist(playlistId = queue.id, episodeId = moved.id, to = finalIndex)) {
            mutationEpoch += 1
            val list = episodes.filterNot { it.id == moved.id }.toMutableList()
            list.add(finalIndex.coerceIn(0, list.size), moved)
            episodes = list
            syncMembership()
        }
    }

    fun moveToTop(episode: EpisodeData) {
        val idx = episodes.indexOfFirst { it.id == episode.id }
        if (idx < 0) return
        move(fromIndex = idx, toIndex = 0)
    }

    /// Called from other tabs' episode rows ("Add to Queue"). Lands at the
    /// FRONT (position 0, newest first) by default — the web's
    /// add_to_queue_front pref, applied optimistically and in the drained op.
    fun add(episode: EpisodeData) {
        val queue = queue ?: run {
            ToastCenter.error("Queue isn't loaded yet. Reconnect once, then retry.")
            return
        }
        if (episodes.any { it.id == episode.id }) return
        val front = core.models?.prefs?.prefs?.addToQueueFront ?: true
        core.enqueueMutation(OutboxOp.Kind.AddToPlaylist(playlistId = queue.id, episodeId = episode.id, position = if (front) 0 else null), episode = episode) {
            mutationEpoch += 1
            if (episodes.none { it.id == episode.id }) episodes = if (front) listOf(episode) + episodes else episodes + episode
            syncMembership()
        }
    }

    /// The rows the view renders. Reordering is only meaningful over the raw
    /// position order with no search/filter applied.
    val displayed: List<EpisodeData>
        get() = query.apply(
            episodes,
            isOnDevice = { id ->
                core.models?.device?.stateOf(id) == DeviceDownloads.State.Downloaded
            },
            status = { core.overlayStatus(it) },
        )

    val reorderable: Boolean
        get() = query == defaultQuery

    fun contains(episode: EpisodeData): Boolean = episodes.any { it.id == episode.id }

    private fun syncMembership() {
        val queue = queue ?: return
        // Keep the playlists pool's episode_ids for the queue row in step
        // (menus and counts read it — see PlaylistsModel.setMembership).
        core.models?.playlists?.setMembership(
            playlistId = queue.id, episodeIds = episodes.map { it.id })

    }

    private companion object {
        const val QUERY_KEY = "listquery-queue"
        val defaultQuery =
            ListQuery(orderField = EpisodeOrderField.Position, direction = OrderDirection.Asc)
    }
}

File: ios/Halogen/Features/Queue/QueueModel.swift
import Foundation
import Observation

/// The Queue tab. The queue IS the user's default playlist (`is_default`) —
/// never a separate concept. Local-first: cached queue renders first, and
/// mutations apply optimistically then sync via the Outbox (fully offline).
@MainActor
@Observable
final class QueueModel {
    private let accountStore: LocalStore?

    private unowned let core: HalogenCore

    private(set) var queue: PlaylistData?
    private(set) var episodes: [EpisodeData] = []
    /// Search + facet + order, applied LOCALLY (the queue is a small,
    /// position-ordered list already in memory). Persisted across relaunch
    /// (web use_list_view_state("queue")).
    var query = ListQuery(orderField: .position, direction: .asc) {
        didSet {
            guard query != oldValue else { return }
            let snapshot = query
            Task { [store = accountStore] in await store?.save(snapshot, key: "listquery-queue") }
        }
    }
    private var loadedQuery = false
    private(set) var loaded = false
    private(set) var error: String?
    /// Bumped synchronously by every optimistic mutation. A refresh captures
    /// it before its fetch and discards the response if it moved — a mutation
    /// landing mid-fetch must never be overwritten by the stale server list
    /// (web: cache_playlists applies its membership guard at CACHE time).
    private var mutationEpoch = 0

    init(core: HalogenCore) {
        self.core = core
        self.accountStore = core.store
    }

    func load() async {
        error = nil
        if !loadedQuery {
            loadedQuery = true
            if let store = accountStore,
                let saved = await store.load(ListQuery.self, key: "listquery-queue")
            {
                query = saved
            }
        }
        if let store = accountStore {
            if queue == nil, let cachedQueue = await store.load(PlaylistData.self, key: CacheKey.queueMeta) {
                queue = cachedQueue
            }
            if let queue, episodes.isEmpty,
                let cached = await store.load(
                    [EpisodeData].self, key: CacheKey.playlistEpisodes(queue.id))
            {
                episodes = cached
                loaded = true
            }
        }
        await refresh()
    }

    func refresh() async {
        // Drain BEFORE pulling (the web's hard-coded order everywhere): the
        // server must reflect optimistic queue membership before we read the
        // playlist back, or this fetch wipes a not-yet-shipped offline add.
        await core.outbox?.drain()
        do {
            let fresh = try await core.forAccount(accountStore).defaultPlaylist()
            queue = fresh
            if let fresh {
                await accountStore?.save(fresh, key: CacheKey.queueMeta)
                // Membership ops still queued for this playlist (the drain
                // couldn't ship them): the local list is AHEAD of the server —
                // keep it, don't overwrite screen/cache with stale membership
                // (web: cache_playlists' membership-preservation invariant).
                if let outbox = core.outbox, await outbox.hasPendingOps(playlistId: fresh.id) {
                    loaded = true
                    return
                }
                let epoch = mutationEpoch
                let served = try await core.forAccount(accountStore).playlistEpisodes(playlistId: fresh.id)
                // Guard AGAIN at cache time: a mutation (or its just-enqueued
                // op) that landed while the fetch was in flight outranks the
                // stale server membership.
                if epoch != mutationEpoch { loaded = true; return }
                if let outbox = core.outbox, await outbox.hasPendingOps(playlistId: fresh.id) {
                    loaded = true
                    return
                }
                episodes = served
                await accountStore?.save(episodes, key: CacheKey.playlistEpisodes(fresh.id))
            } else {
                episodes = []
                await accountStore?.remove(key: CacheKey.queueMeta)
            }
            error = nil
        } catch {
            if episodes.isEmpty && queue == nil { self.error = FriendlyError.message(error) }
        }
        loaded = true
    }

    /// "No queue yet" flow: create the default playlist (online-only — the
    /// created id anchors every later offline op). Failure toasts — setting
    /// `error` replaced the whole no-queue screen (and its Create button)
    /// with a full-page error.
    func createQueue() async {
        do {
            _ = try await core.forAccount(accountStore).createPlaylist(name: "Queue", isDefault: true)
            await refresh()
        } catch {
            ToastCenter.shared.error(
                "Couldn't create the queue — \(FriendlyError.message(error))")
        }
    }

    // MARK: - optimistic membership ops (offline-capable via Outbox)

    func remove(_ episode: EpisodeData) {
        guard let queue else { return }
        core.enqueueMutation(
            originStore: accountStore, .removeFromPlaylist(playlistId: queue.id, episodeId: episode.id)
        ) { [self] in
            mutationEpoch += 1
            episodes.removeAll { $0.id == episode.id }
            persistSnapshot()
        }
    }

    func move(fromOffsets: IndexSet, toOffset: Int) {
        guard let fromIndex = fromOffsets.first, episodes.indices.contains(fromIndex) else { return }
        let episode = episodes[fromIndex]
        var ordered = episodes
        ordered.move(fromOffsets: fromOffsets, toOffset: toOffset)
        guard let destination = ordered.firstIndex(where: { $0.id == episode.id }) else { return }
        move(episode, to: destination)
    }

    private func move(_ episode: EpisodeData, to destination: Int) {
        guard let queue else { return }
        let operation = OutboxOp.Kind.moveInPlaylist(
            playlistId: queue.id, episodeId: episode.id, to: Int32(destination))
        core.enqueueMutation(originStore: accountStore, operation) { [self] in
            guard let index = episodes.firstIndex(where: { $0.id == episode.id }) else { return }
            mutationEpoch += 1
            let moved = episodes.remove(at: index)
            episodes.insert(moved, at: min(destination, episodes.count))
            persistSnapshot()
        }
    }

    func moveToTop(_ episode: EpisodeData) { move(episode, to: 0) }

    func add(_ episode: EpisodeData) {
        guard let queue else {
            ToastCenter.shared.error("Queue isn't loaded yet. Reconnect once, then retry.")
            return
        }
        let front = core.models?.prefs.prefs.addToQueueFront ?? true
        core.enqueueMutation(
            originStore: accountStore,
            .addToPlaylist(playlistId: queue.id, episodeId: episode.id, position: front ? 0 : nil), episode: episode
        ) { [self] in
            guard !episodes.contains(where: { $0.id == episode.id }) else { return }
            mutationEpoch += 1
            if front { episodes.insert(episode, at: 0) } else { episodes.append(episode) }
            persistSnapshot()
        }
    }

    /// The rows the view renders. Reordering is only meaningful over the raw
    /// position order with no search/filter applied.
    var displayed: [EpisodeData] {
        query.apply(
            to: episodes,
            isOnDevice: { [weak core] id in core?.models?.device.state(of: id) == .downloaded },
            status: { [weak core] in core?.overlayStatus($0) ?? $0.playback_status ?? .unplayed }
        )
    }

    var reorderable: Bool {
        query == ListQuery(orderField: .position, direction: .asc)
    }

    func contains(_ episode: EpisodeData) -> Bool {
        episodes.contains(where: { $0.id == episode.id })
    }

    private func persistSnapshot() {
        guard let queue else { return }
        // Keep the playlists pool's episode_ids for the queue row in step
        // (menus and counts read it — see PlaylistsModel.setMembership).
        core.models?.playlists.setMembership(
            playlistId: queue.id, episodeIds: episodes.map(\.id))

    }
}
